pkg/runtime/gomemlimit_linux.go

```diff
@@ -3,49 +3,141 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build linux && go1.19
+//go:build linux
 
 package runtime
 
 import (
+	"context"
 	"errors"
-	"os"
+	"math"
 	"runtime/debug"
+	"time"
 
 	"github.com/DataDog/datadog-agent/pkg/util/cgroups"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-// SetGoMemLimit configures Go memory soft limit based on cgroups.
-// The soft limit is set to 90% of the cgroup memory hard limit.
-// The function is noop if
-//   - GOMEMLIMIT is set already
-//   - There is no cgroup limit
-//
-// Read more about Go memory limit in https://tip.golang.org/doc/gc-guide#Memory_limit
-func SetGoMemLimit(isContainerized bool) error {
-	if _, ok := os.LookupEnv("GOMEMLIMIT"); ok {
+type staticMemoryLimiter struct {
+	limitPct        float64
+	isContainerized bool
+}
+
+// NewStaticMemoryLimiter return a memory limit that sets memory limit once based on cgroup limit
+func NewStaticMemoryLimiter(limitPct float64, isContainerized bool) MemoryLimiter {
+	return staticMemoryLimiter{
+		limitPct:        limitPct,
+		isContainerized: isContainerized,
+	}
+}
+
+// Run runs the memory limit logic
+func (s staticMemoryLimiter) Run(context.Context) error {
+	// SetMemoryLimit with negative values returns current value
+	if debug.SetMemoryLimit(-1) != math.MaxInt64 {
 		log.Debug("GOMEMLIMIT is set already, doing nothing")
 		return nil
 	}
-	selfReader, err := cgroups.NewSelfReader("/proc", isContainerized)
+
+	selfCgroup, err := getSelfCgroup(s.isContainerized)
```

Comment:
What if we're not running in a container? This return an `err` and we don't use GOMEMLIMIT? Or even if not containerised the Agent is part of a cgroup and will have something usable?

Reply:
It works if the Agent is simply in a dedicated `cgroup` (like systemd managed units), it's just that we need this parameter to accurately get self limit.